amaranth_cli/__init__.py

```diff
@@ -0,0 +1,130 @@
+"""
+This file is not a part of the Amaranth module tree because the CLI needs to emit Make-style
+dependency files as a part of the generation process. In order for `from amaranth import *`
+to work as a prelude, it has to load several of the files under `amaranth/`, which means
+these will not be loaded later in the process, and not recorded as dependencies.
+"""
+
+import importlib
+import argparse
+import stat
+import sys
+import os
+import re
+
+
+def _build_parser():
+    def component(reference):
+        from amaranth import Elaboratable
+
+        if m := re.match(r"(\w+(?:\.\w+)*):(\w+(?:\.\w+)*)", reference, re.IGNORECASE|re.ASCII):
+            mod_name, qual_name = m[1], m[2]
+            try:
+                obj = importlib.import_module(mod_name)
+            except ImportError as e:
+                raise argparse.ArgumentTypeError(f"{mod_name!r} does not refer to "
+                                 "an importable Python module") from e
+            try:
+                for attr in qual_name.split("."):
+                    obj = getattr(obj, attr)
+            except AttributeError as e:
+                raise argparse.ArgumentTypeError(f"{qual_name!r} does not refer to an object "
+                                 f"within the {mod_name!r} module") from e
+            if not issubclass(obj, Elaboratable):
+                raise argparse.ArgumentTypeError(f"'{qual_name}:{mod_name}' refers to an object that is not elaboratable")
+            return obj
+        else:
+            raise argparse.ArgumentTypeError(f"{reference!r} is not a Python object reference")
```

Comment:
It took me a while to realize that by "object reference", the error means "your input argument fails to match a regex", rather than "Python couldn't find an object with that name" (which is what the `ImportError` exception handler is for).

Perhaps this `else` branch could have "expected format is `path.to.mod:Obj`" or something?

Reply:
Yep this could be improved!!

Reply:
Care to submit a PR against a PR?

Reply:
Will do in the coming days (my response time will be sporadic until possibly this Monday afternoon).

Reply:
PR is done: https://github.com/whitequark/amaranth/pull/1